src/eterna/folding/Folder.ts

```diff
@@ -19,13 +20,16 @@ export interface FullEvalCache {
 }
 
 export interface SuboptEnsembleResult {
+    suboptStructures: string[];
+    suboptEnergyError: number[];
+    suboptFreeEnergy: number[];
+}
+
+export interface DefectResult {
     ensembleDefect: number;
     ensembleDefectNormalized: number;
     mfeDefect: number;
     mfeDefectNormalized: number;
```

Comment:
Just to ask the question (as I don't thoroughly understand defect): Should these always be computed together? Or should we split these out even further and only calculate what the user asks for? Particularly the "normalized" versions seem trivial to compute and are straightforwardly derived from the "normal" defect - there's an argument to be had for having the caller (eg, the Eternascript using the values) compute these if they're needed.

Also as a note, looking at the defect code, I'm wondering if more of this should just be done in TS to begin with. In general, I think we want to only do the minimum amount necessary in C++ (where performance matters or we actually have to call C/C++ functions), plus it looks like you're re-computing things like MFE and energy which we may already have cached in TS.

Reply:
the call for FullEnsembleDefect inFullEnsemble.cpp has a variable called "mode" that is a integer that tells the function to find normal=1, mfe=2, or both=3. The external interface DefectResult  right now is coded up to return a 0 or -1 or a particular defect value if you did not ask for it / it was not computed. 

good point. there are a couple calls for structures and MFE. If we want to shift to more being done in TS and less in C++ then I can just make sure the nupack calls we need are exposed and then do those calls or grab from cache if available for that call in the future. You also make a good point about seperating out the ensemble defect portions out.

Reply:
I see some places were I can easily do some stuff in TS and it is harder in in others.What I intend to do at a minimum is to write it so that in TS it checks for a cached MFE struct and if available it will pass that to this function. This implementation then naturally leads to a separate ensembleDefect result and a mfe defect result in TS per @luxaritas previous comment about  further breaking up ED and MFE_ED results. As such I will break out ED and MFE_ED into separate calls and functions